### PR TITLE
Complex log accuracy refinement

### PR DESCRIPTION
**Pull request by s-oboyle**
## Updated the complex log function to avoid under/overflow issues.

There are a few issues with the bounds in the clog function relating to under/overflow.

In the current version, it can be the case that the real part of the answer, which is implemented as:
`(log(abs(__x)), arg(__x));)`,

can have `abs(__x)` under or overflow, while `log(abs(__x))` is finite.

Of more importance when calculating with non-extreme values is that it also suffers catastrophic cancellation around `abs(__x) ~= 1.0`, where the loss of accuracy around `1.0` when calculating `abs(x)` gets passed to `log()` and gives a large relative/ulp error.

This new version fixes these issues, and through inlining some function calls does so while retaining as much perf as possible.

The combination of this MR and the [cexp improvement MR](https://github.com/NVIDIA/cccl/pull/4928) has a knock on effect on the cpow function, which are now generally more accurate and support more inputs (although could still do with their own refresh).

Also added a quick improvement for `clog10`, changing a divide that the compiler could not optimize to a multiply.

## Perf
For fp64 precision the performance was faster on everything.
For fp32 precision the performance was slower on everything.
The jump in accuracy from the fp32 current to new implementation is a larger than the accuracy jump in the fp64 version, perhaps this goes some way in explaining the diff.

Using the math-teams standard `math_bench` test we have the following on H100 (averaged summary):
**Operations/SM/cycle:**
`clog():`
| H100 |  old | new | new/old |
| --- | --- | --- | --- |
| fp64 | 0.41452 | 0.42906 | 1.035 | 
| fp32 | 1.18094 | 1.05512 | 0.893 |

`clog10():`
| H100 |  old | new | new/old |
| --- | --- | --- | --- |
| fp64 | 0.30676 | 0.42044 | 1.370| 
| fp32 | 0.9704 | 1.04182 | 1.073 |

`pow():`
| H100 |  old | new | new/old |
| --- | --- | --- | --- |
| fp64 | 0.20486 | 0.22584 | 1.102 | 
| fp32 | 0.62774 | 0.58364 | 0.929 |

(See https://docs.google.com/spreadsheets/d/1TdOpEbLgoL1QOWKjeO3pwFDDMoPr4iqnPjz_XrL7UEA for raw/non-averaged info. NV internal.)

## Correctness GPU/CPU
With the current function, as well as having issues whenever `real^2 + imag^2` under or overflows, suffers from bad accuracy loss around `real^2 + imag^2 = ~1.0`

Eg: A quick fp64 search yields a relative error of 0.0001731, or ulp of 1.028e12:
```
Max ulp real error (1.028e+12,0.02957) @ (1,3.794584201e-07)    (0x3ff0000000000402,0x3e99770bc7cde226)
        Ours = (2.997602166e-13,3.794584201e-07)    Ref = (2.998121109e-13,3.794584201e-07)
        Ours = (0x3d5517fffffffc86,0x3e99770bc7cdda6d)               Ref = (0x3d5518ef51f4ffc4,0x3e99770bc7cdda6d)
```
while in fp32 it gets much worse, here with a relative error of 2.999, or a ulp of 2.516e7:
```
Max ulp real error (2.516e+07,0.06174) @ (0.008398859762,0.9999647141)  (0x3c099b5f,0x3f7ffdb0)
        Ours = (-5.960464478e-08,1.562397361)    Ref = (-1.490450963e-08,1.562397361)
        Ours = (0xb3800000,0x3fc7fca3)               Ref = (0xb280075d,0x3fc7fca3)
```

For the new version, an intensive bracket and bisect search, along with testing special hard values, gives:
# GPU
```
GPU fp64:
Max ulp real error (2.275,0.2758) @ (-4.243991582e-314,1.251714353)     (0x8000000200000000,0x3ff40705a109caa0)
        Ours = (0.2245140938,1.570796327)    Ref = (0.2245140938,1.570796327)
        Ours = (0x3fccbce0b9397b9c,0x3ff921fb54442d18)               Ref = (0x3fccbce0b9397b9e,0x3ff921fb54442d18)

Max ulp imag error (0.513,1.688) @ (4.400651191,-5.855202676)   (0x40119a444e4e2032,0xc0176bba40116ebb)
        Ours = (1.99123262,-0.9262850069)    Ref = (1.99123262,-0.9262850069)
        Ours = (0x3fffdc16bc4406e3,0xbfeda42074669ed7)               Ref = (0x3fffdc16bc4406e4,0xbfeda42074669ed9)
```

```
GPU fp32:
Max ulp real error (2.119,0.5114) @ (0.8844045997,-0.001042612363)      (0x3f626857,0xba88a844)
        Ours = (-0.12283995,-0.001178885577)    Ref = (-0.1228399351,-0.001178885694)
        Ours = (0xbdfb9383,0xba9a84d6)               Ref = (0xbdfb9381,0xba9a84d7)

Max ulp imag error (0.2917,1.816) @ (8.296953091e-35,-2.090229989e-35)  (0x6dc9223,0x85de4579)
        Ours = (-78.44382477,-0.2467918545)    Ref = (-78.44382477,-0.2467918843)
        Ours = (0xc29ce33d,0xbe7cb701)               Ref = (0xc29ce33d,0xbe7cb703)
```

```
GPU fp16:
Max ulp real error (0.5002,0.4331) @ (-0.006740570068,0.9760742188)     (0x9ee7,0x3bcf)
        Ours = (-0.02420043945,1.578125)    Ref = (-0.02418518066,1.578125)
        Ours = (0xa632,0x3e50)               Ref = (0xa631,0x3e50)

Max ulp imag error (0.2325,0.5001) @ (713,264)  (0x6192,0x5c20)
        Ours = (6.6328125,0.3544921875)    Ref = (6.6328125,0.3547363281)
        Ours = (0x46a2,0x35ac)               Ref = (0x46a2,0x35ad)
```

```
GPU bf16:
Max ulp real error (0.5,0.4744) @ (3.771741203e-30,2.388153131e-32)     (0xe99,0xaf8)
        Ours = (-68,0.006317138672)    Ref = (-67.5,0.006317138672)
        Ours = (0xc288,0x3bcf)               Ref = (0xc287,0x3bcf)

Max ulp imag error (0.3976,0.5) @ (0.02990722656,-6.8125)       (0x3cf5,0xc0da)
        Ours = (1.921875,-1.5625)    Ref = (1.921875,-1.5703125)
        Ours = (0x3ff6,0xbfc8)               Ref = (0x3ff6,0xbfc9)
```

# CPU
```
CPU fp64:
Max ulp real error (2.27,0.2758) @ (-0,-1.251099866)    (0x8000000000000000,0xbff404814aee9000)
        Ours = (0.2240230569,-1.570796327)    Ref = (0.2240230569,-1.570796327)
        Ours = (0x3fccacc99b8d8178,0xbff921fb54442d18)               Ref = (0x3fccacc99b8d817a,0xbff921fb54442d18)

Max ulp imag error (0.07374,0.521) @ (27.69782245,-3.198612825) (0x403bb2a47deecc00,0xc00996c252265800)
        Ours = (3.327977826,-0.1149731688)    Ref = (3.327977826,-0.1149731688)
        Ours = (0x400a9fb2d69f6014,0xbfbd6ee1b00e2e40)               Ref = (0x400a9fb2d69f6014,0xbfbd6ee1b00e2e41)
```

```
CPU fp32:
Max ulp real error (2.06,0.4104) @ (-0.8834761977,9.203773743e-09)      (0xbf622b7f,0x321e1ea0)
        Ours = (-0.1238909438,3.141592741)    Ref = (-0.1238909289,3.141592741)
        Ours = (0xbdfdba89,0x40490fdb)               Ref = (0xbdfdba87,0x40490fdb)

Max ulp imag error (0.06559,1.478) @ (27.61241913,3.469610214)  (0x41dce63c,0x405e0e18)
        Ours = (3.326098442,0.1249988452)    Ref = (3.326098442,0.1249988526)
        Ours = (0x4054decc,0x3dffff65)               Ref = (0x4054decc,0x3dffff66)
```

```
CPU fp16:
Max ulp real error (0.5002,0.2219) @ (0.7919921875,0.1275634766)        (0x3a56,0x3015)
        Ours = (-0.2204589844,0.1596679688)    Ref = (-0.2203369141,0.1596679688)
        Ours = (0xb30e,0x311c)               Ref = (0xb30d,0x311c)

Max ulp imag error (0.04422,0.5001) @ (57152,70.9375)   (0x7afa,0x546f)
        Ours = (10.953125,0.00124168396)    Ref = (10.953125,0.001240730286)
        Ours = (0x497a,0x1516)               Ref = (0x497a,0x1515)
```

```
CPU bf16: 
Max ulp real error (0.5,0.4744) @ (3.771741203e-30,2.388153131e-32)     (0xe99,0xaf8)
        Ours = (-68,0.006317138672)    Ref = (-67.5,0.006317138672)
        Ours = (0xc288,0x3bcf)               Ref = (0xc287,0x3bcf)

Max ulp imag error (0.4747,0.5) @ (4.512836137e+32,3.468292042e+33)     (0x75b2,0x772b)
        Ours = (77,1.4375)    Ref = (77,1.4453125)
        Ours = (0x429a,0x3fb8)               Ref = (0x429a,0x3fb9)
```

## Note on `clog10`:
The accuracy of the current `clog10` function is about the same as the current `clog` function.

With the new `clog` function, but with it still dividing by a constant it reduced down to ~5ulp's, and changing the divide to a constant multiply gives ~3ulps, or what you would expect multiplying `clog` by an irrational number rounded to an fp-type.

### CODE REVIEW

**Comment by s-oboyle** [comment]
@miscco another one for you.

**Comment by s-oboyle** [comment]
/ok to test [fbcfb1a](https://github.com/NVIDIA/cccl/pull/5185/commits/fbcfb1ae935fb0e03b15ce9956118f6403bbaa13)

File: libcudacxx/include/cuda/std/__complex/logarithms.h [hunk 1]
```diff
@@ -23,28 +23,277 @@
...
+  constexpr int32_t __mant_nbits = __fp_mant_nbits_v<__fp_format_of_v<_Tp>>;
+  constexpr int32_t __exp_bias   = __fp_exp_bias_v<__fp_format_of_v<_Tp>>;
+
+  // Cut off the hi and low bit of __exp_mask.
+  // 0x3F000000 for fp32,
+  // 0x3FE0000000000000 for fp64 etc
+  constexpr _UINT_T __exp_mask_of_half = ((__exp_mask >> (__mant_nbits + 2)) << (__mant_nbits + 1));
+
+  _CONST_TP __real_abs = _CUDA_VSTD::fabs(__x.real());
+  _CONST_TP __imag_abs = _CUDA_VSTD::fabs(__x.imag());
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__complex/logarithms.h]: This should just be
```suggestion
  const _Tp __real_abs = _CUDA_VSTD::fabs(__x.real());
  const _Tp __imag_abs = _CUDA_VSTD::fabs(__x.imag());
```

File: libcudacxx/include/cuda/std/__complex/logarithms.h [hunk 2]
```diff
@@ -23,28 +23,277 @@
...
+  return __log1p_poly;
+}
+
 template <class _Tp>
 [[nodiscard]] _CCCL_API inline complex<_Tp> log(const complex<_Tp>& __x)
 {
-  return complex<_Tp>(_CUDA_VSTD::log(_CUDA_VSTD::abs(__x)), _CUDA_VSTD::arg(__x));
+  // Uint of the same size as our fp type.
+  // Shouldn't need make_unsigned, but just in case:
+  using _UINT_T   = make_unsigned_t<__fp_storage_of_t<_Tp>>;
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__complex/logarithms.h]: We conventionally try to use `_UInt_t` or `__uint_t` or maybe even better just `__storage_t`

We are really unconsistent with aliases, but do try to reserve SHOUTY_NAMES to macros

File: libcudacxx/include/cuda/std/__complex/logarithms.h [hunk 3]
```diff
@@ -23,28 +23,277 @@
...
+  // The below code inlines and removes these two calls:
+  //    _Tp __max_reduced = _CUDA_VSTD::frexpf(__max, &__exp);
+  //    _Tp __min_reduced = _CUDA_VSTD::ldexpf(__min, -__exp);
+
+  // Unbiased exponent of 2.0*__max
+  int32_t __exp = static_cast<int32_t>(reinterpret_cast<_UINT_T&>(__max) >> __mant_nbits) - __exp_bias + 1;
+
+  // Quick frexp:
+  _UINT_T __max_reduced_as_uint = (reinterpret_cast<_UINT_T&>(__max) & __mant_mask) | __exp_mask_of_half;
+  _Tp __max_reduced             = reinterpret_cast<_Tp&>(__max_reduced_as_uint);
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__complex/logarithms.h]: We should rather use `_CUDA_VSTD::bit_cast<__storage_t>(__max)`

That will allow us to actually make this function `constexpr` for most of the supported compilers

**Review by miscco — COMMENTED** [review]
Thanks a lot for working on this, much appreciated.

Some general advice regarding use of `bit_cast` / naming

File: libcudacxx/include/cuda/std/__complex/logarithms.h (hunk 1 above)
Comment (s-oboyle) [reply libcudacxx/include/cuda/std/__complex/logarithms.h]: Done

File: libcudacxx/include/cuda/std/__complex/logarithms.h (hunk 2 above)
Comment (s-oboyle) [reply libcudacxx/include/cuda/std/__complex/logarithms.h]: I might leave it as `__uint_t` if that's ok, it will make it a bit easier for whatever math person ends up looking at this again. Figuring out the `__storage_t` system took me some time.

File: libcudacxx/include/cuda/std/__complex/logarithms.h (hunk 3 above)
Comment (s-oboyle) [reply libcudacxx/include/cuda/std/__complex/logarithms.h]: Yes, thank you I forgot that was a thing. Changed.

**Comment by s-oboyle** [comment]
@miscco I've had to modify the tests a little, the `__half` and `__nv_bfloat16` pow functions are now too accurate for the `exp(y * log(x))` reference implementation.
The `log10` test might also fail for these types now, as they are again more accurate, but only by a little bit now the test might just pass.

File: libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/pow_complex_complex.pass.cpp [hunk 4]
```diff
@@ -37,6 +37,23 @@ __host__ __device__ void test()
...
+  }
+#endif // _LIBCUDACXX_HAS_NVFP16()
+
+#if _LIBCUDACXX_HAS_NVBF16()
+  if (cuda::std::is_same<T, __nv_bfloat16>)
+  {
+    return;
+  }
+#endif // _LIBCUDACXX_HAS_NVBf16()
+
```
Comment (miscco) [inline libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/pow_complex_complex.pass.cpp]: I'd rather have we skip the accuracy check below and test the edge cases than not testing those at all

Comment (s-oboyle) [reply libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/pow_complex_complex.pass.cpp]: Ok, I've modified that in.
I'm a slightly wary that it is easy to generate incorrect NaN outputs from `pow`, (and in general `pow` is just a tricky function, especially the complex version), so some inputs might get detected in the NaN corner case check as opposed to an accuracy check.
I haven't looked at the exact inputs, but I guess we can run it and see if this is an issue or not.

**Comment by s-oboyle** [comment]
/ok to test [beb306b](https://github.com/NVIDIA/cccl/pull/5185/commits/beb306bcf617b4b06e75e6bbd9e69ee02bd11b7d)

**Comment by s-oboyle** [comment]
/ok to test [beb306b](https://github.com/NVIDIA/cccl/pull/5185/commits/beb306bcf617b4b06e75e6bbd9e69ee02bd11b7d)

**Comment by s-oboyle** [comment]
/ok to test [bcee857](https://github.com/NVIDIA/cccl/pull/5185/commits/bcee8571ebd2648240b3b6173c83854d059dbf92)

File: libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/log10.pass.cpp [hunk 5]
```diff
@@ -39,7 +39,19 @@ __host__ __device__ void test_edges()
...
+
+    // The __half or __nv_float16 functions use fp32, we need to account for this
+    // as we are checking for floating-point equality:
+#if _LIBCUDACXX_HAS_NVFP16()
+    z = log<float>(cuda::std::complex<float>(testcases[i])) * 0.434294481903251827651128918916605082294397f;
+#endif // _LIBCUDACXX_HAS_NVFP16()
+
+#if _LIBCUDACXX_HAS_NVBF16()
+    z = log<float>(cuda::std::complex<float>(testcases[i])) * 0.434294481903251827651128918916605082294397f;
+#endif // _LIBCUDACXX_HAS_NVBF16()
```
Comment (miscco) [inline libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/log10.pass.cpp]: I believe this is not valid, because you are applying this twice if both `__half` and `__nvbfloat16` are available

This would need to be an 
`if constexpr(_CUDA_VSTD::__is_extended_flaoting_point_v<T>)`

Comment (s-oboyle) [reply libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/log10.pass.cpp]: Well spotted, yes I somehow deleted the `is_same()` check from the commit...

**Comment by s-oboyle** [comment]
/ok to test [ff6fec5](https://github.com/NVIDIA/cccl/pull/5185/commits/ff6fec5154e81cfd74ecc9dec3d4b85b0a1af33a)

File: libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/log10.pass.cpp [hunk 6]
```diff
@@ -45,11 +45,17 @@ __host__ __device__ void test_edges()
...
+    }
 #endif // _LIBCUDACXX_HAS_NVFP16()
 
 #if _LIBCUDACXX_HAS_NVBF16()
-    z = log<float>(cuda::std::complex<float>(testcases[i])) * 0.434294481903251827651128918916605082294397f;
+    if (cuda::std::is_same<T, __nv_bfloat16>::value)
+    {
+      z = log<float>(cuda::std::complex<float>(testcases[i])) * 0.434294481903251827651128918916605082294397f;
+    }
 #endif // _LIBCUDACXX_HAS_NVBF16()
```
Comment (miscco) [inline libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/log10.pass.cpp]: Both conditions are the same so we can just simplify to 
```suggestion
    if constexpr (cuda::std::__is_extended_floating_point_v<T>)
    {
      z = cuda::std::complex<T>{_CUDA_VSTD::log(cuda::std::complex<float>{testcases[i])} * cuda::std::log10e_v<float>};
    }
```

Comment (miscco) [reply libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/log10.pass.cpp]: Also note that we should use `if constexpr` for compile time switches

File: libcudacxx/include/cuda/std/__complex/logarithms.h [hunk 7]
```diff
@@ -23,28 +23,276 @@
...
+}
+#endif // _LIBCUDACXX_HAS_NVFP16()
+
 // log10
 
 template <class _Tp>
 [[nodiscard]] _CCCL_API inline complex<_Tp> log10(const complex<_Tp>& __x)
 {
-  return _CUDA_VSTD::log(__x) / _CUDA_VSTD::log(_Tp(10));
+  return _CUDA_VSTD::log(__x) * _Tp(_CUDA_VSTD::numbers::log10e_v<_Tp>);
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__complex/logarithms.h]: We should do
```suggestion
  return _CUDA_VSTD::log(__x) * __numbers<_Tp>::__log10e();
```

File: libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/log10.pass.cpp [hunk 8]
```diff
@@ -39,7 +39,25 @@ __host__ __device__ void test_edges()
   for (unsigned i = 0; i < N; ++i)
   {
     cuda::std::complex<T> r = log10(testcases[i]);
-    cuda::std::complex<T> z = log(testcases[i]) / cuda::std::log(T(10.0));
+    const T log10_e         = T(0.434294481903251827651128918916605082294397);
```
Comment (miscco) [inline libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/log10.pass.cpp]: I think we should be fine doing
```suggestion
    const T log10_e         = cuda::std::__numbers<T>::__log10e();
```

The issue here being that MSVC does not allow the inline variables because of ... reasons

File: libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/log10.pass.cpp (hunk 6 above)
Comment (s-oboyle) [reply libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/log10.pass.cpp]: Does `__is_extended_floating_point_v` include types like fp80 or fp128? Would cause confusion in the future.

File: libcudacxx/include/cuda/std/__complex/logarithms.h (hunk 7 above)
Comment (s-oboyle) [reply libcudacxx/include/cuda/std/__complex/logarithms.h]: So I had it like this before but I was getting `log10e_v is not defined` errors.
Even from just the last test run I'm getting:
```C:/cccl/libcudacxx/include\cuda/std/__complex/logarithms.h(278): error: identifier "cuda::std::__4::numbers::log10e_v<float> " is undefined in device code```
for a few arhcitectures.

Comment (s-oboyle) [reply libcudacxx/include/cuda/std/__complex/logarithms.h]: It also somewhat bothers me that doing it this way is just comparing the function implementation against itself basically, if someone/something changes the value of `log10e_v` this test will not fail etc.

File: libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/log10.pass.cpp (hunk 6 above)
Comment (miscco) [reply libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/log10.pass.cpp]: It does include all NVIDIA extended floating point types, which probably are emulated through float for things that are not specialized in libdevice

File: libcudacxx/include/cuda/std/__complex/logarithms.h (hunk 7 above)
Comment (miscco) [reply libcudacxx/include/cuda/std/__complex/logarithms.h]: fine by me

File: libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/log10.pass.cpp (hunk 6 above)
Comment (s-oboyle) [reply libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/log10.pass.cpp]: I guess that's ok for now.

File: libcudacxx/include/cuda/std/__complex/logarithms.h (hunk 7 above)
Comment (s-oboyle) [reply libcudacxx/include/cuda/std/__complex/logarithms.h]: That should be fine, changed.

File: libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/log10.pass.cpp [hunk 9]
```diff
@@ -39,7 +39,25 @@ __host__ __device__ void test_edges()
...
   {
     cuda::std::complex<T> r = log10(testcases[i]);
-    cuda::std::complex<T> z = log(testcases[i]) / cuda::std::log(T(10.0));
+    const T log10_e         = cuda::std::__numbers<T>::__log10e();
+    cuda::std::complex<T> z = log(testcases[i]) * log10_e;
+
+    // The __half or __nv_float16 functions use fp32, we need to account for this
+    // as we are checking for floating-point equality:
+#if _LIBCUDACXX_HAS_NVFP16()
+    if (cuda::std::is_same<T, __half>::value)
```
Comment (miscco) [inline libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/log10.pass.cpp]: This should be 
```suggestion
    if constexpr (cuda::std::is_same_v<T, __half>)
```

File: libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/log10.pass.cpp [hunk 10]
```diff
@@ -39,7 +39,25 @@ __host__ __device__ void test_edges()
...
+    {
+      z = log<float>(cuda::std::complex<float>(testcases[i])) * 0.434294481903251827651128918916605082294397f;
+    }
+#endif // _LIBCUDACXX_HAS_NVFP16()
+
+#if _LIBCUDACXX_HAS_NVBF16()
+    if (cuda::std::is_same<T, __nv_bfloat16>::value)
```
Comment (miscco) [inline libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/log10.pass.cpp]: Ditto
```suggestion
    if constexpr (cuda::std::is_same_v<T, __nv_bfloat16>)
```

File: libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/pow_complex_complex.pass.cpp [hunk 11]
```diff
@@ -44,7 +44,26 @@ __host__ __device__ void test_edges()
     for (unsigned j = 0; j < N; ++j)
     {
       cuda::std::complex<T> r = cuda::std::pow(testcases[i], testcases[j]);
-      cuda::std::complex<T> z = cuda::std::exp(testcases[j] * log(testcases[i]));
+      cuda::std::complex<T> z = cuda::std::exp(testcases[j] * cuda::std::log(testcases[i]));
+
+      // The __half or __nv_float16 functions use fp32, we need to account for this
+      // as we are checking for floating-point equality:
+#if _LIBCUDACXX_HAS_NVFP16()
+      if (cuda::std::is_same<T, __half>::value)
```
Comment (miscco) [inline libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/pow_complex_complex.pass.cpp]:
```suggestion
      if constexpr (cuda::std::is_same_v<T, __half>)
```

File: libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/pow_complex_complex.pass.cpp [hunk 12]
```diff
@@ -44,7 +44,26 @@ __host__ __device__ void test_edges()
...
+      {
+        z = cuda::std::exp<float>(
+          cuda::std::complex<float>(testcases[j]) * cuda::std::log<float>(cuda::std::complex<float>(testcases[i])));
+      }
+#endif // _LIBCUDACXX_HAS_NVFP16()
+
+#if _LIBCUDACXX_HAS_NVBF16()
+      if (cuda::std::is_same<T, __nv_bfloat16>::value)
```
Comment (miscco) [inline libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/pow_complex_complex.pass.cpp]:
```suggestion
      if constexpr (cuda::std::is_same_v<T, __nv_bfloat16>)
```

File: libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/pow_scalar_complex.pass.cpp [hunk 13]
```diff
@@ -44,7 +44,27 @@ __host__ __device__ void test_edges()
...
     {
       cuda::std::complex<T> r = pow(real(testcases[i]), testcases[j]);
-      cuda::std::complex<T> z = exp(testcases[j] * log(cuda::std::complex<T>(real(testcases[i]))));
+      cuda::std::complex<T> z =
+        cuda::std::exp(testcases[j] * cuda::std::log(cuda::std::complex<T>(real(testcases[i]))));
+
+      // The __half or __nv_float16 functions use fp32, we need to account for this
+      // as we are checking for floating-point equality:
+#if _LIBCUDACXX_HAS_NVFP16()
+      if (cuda::std::is_same<T, __half>::value)
```
Comment (miscco) [inline libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/pow_scalar_complex.pass.cpp]:
```suggestion
      if constexpr (cuda::std::is_same_v<T, __half>)
```

File: libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/pow_scalar_complex.pass.cpp [hunk 14]
```diff
@@ -44,7 +44,27 @@ __host__ __device__ void test_edges()
...
+      {
+        z = cuda::std::exp<float>(
+          cuda::std::complex<float>(testcases[j]) * cuda::std::log<float>(cuda::std::complex<float>(testcases[i])));
+      }
+#endif // _LIBCUDACXX_HAS_NVFP16()
+
+#if _LIBCUDACXX_HAS_NVBF16()
+      if (cuda::std::is_same<T, __nv_bfloat16>::value)
```
Comment (miscco) [inline libcudacxx/test/libcudacxx/std/numerics/complex.number/complex.transcendentals/pow_scalar_complex.pass.cpp]:
```suggestion
      if constexpr (cuda::std::is_same_v<T, __nv_bfloat16>)
```

**Review by miscco — APPROVED** [review]

**Comment by s-oboyle** [comment]
/ok to fix [033ccf4](https://github.com/NVIDIA/cccl/pull/5185/commits/033ccf4582627e3b4c01d703ac085f84690bd9ff)

**Comment by s-oboyle** [comment]
/ok to test [81d796c](https://github.com/NVIDIA/cccl/pull/5185/commits/81d796ca76bb6e73834f713b2ded12d6d5570b3b)

**Comment by s-oboyle** [comment]
/ok to test [33efed4](https://github.com/NVIDIA/cccl/pull/5185/commits/33efed48899c5d23747512f42d7d9854e1ad0f00)
